### PR TITLE
Stats: Adding skeleton loaders to stats module

### PR DESCRIPTION
**Pull request by grzegorz-cp**
Related to https://github.com/Automattic/red-team/issues/54

## Proposed Changes

* unifying TS types
* adding a skeleton loader for the new components

UI with the new skeleton loaders:
SCR-20240704-meyd

## Why are these changes being made?

* code maintenance and improving user experience

## Testing Instructions

* go to the live branch
* apply `stats/empty-module-traffic` feature flag
* navigate to several blogs and verify that the new components load properly

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by kangzj — COMMENTED** [review]
Thanks for working on this!

The country module works as described; however Referrers and 'Post and Pages' don't - they show the empty states while loading data when selecting a different day.

I also don't find the loading state in the design - I'm happy for it to ship after fixing the loading state for the other two modules, but I guess we should probably ask for reviews for designers.

**Comment by kangzj** [comment]
Forgot to upload the video:

https://github.com/Automattic/wp-calypso/assets/1425433/e3315239-b968-431e-b918-6ce08ef5d777

**Comment by grzegorz-cp** [comment]
Thanks @kangzj, the first two modules don't show the loading states because redux returns and empty array instead of null when the data is missing. p1719806322698149-slack-C04UE0ANHDY
I'll investigate this but it's not a priority for now.

**Review by annacmc — APPROVED** [review]
Nice one! I really like this idea 😄  It's working well for me (other than the issues already flagged). I think it will help a lot with performance and user experience in general.

**Review by annacmc — APPROVED** [review]
Nice one! I really like this idea 😄  It's working well for me (other than the issues already flagged). I think it will help a lot with performance and user experience in general.

**Comment by kangzj** [comment]
Looks like the PR needs some type fixes:

```[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10325_10325&logFilter=debug&logView=flowAware)
  Task type_check_client failed after 67892ms.

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10326_10326&logFilter=debug&logView=flowAware)
       Output for type_check_client

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10327_10327&logFilter=debug&logView=flowAware)
    If no output is shown, look for it under the test section for this task.

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10328_10328&logFilter=debug&logView=flowAware)
    ....STDOUT....

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10329_10329&logFilter=debug&logView=flowAware)
    client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx(26,2): error TS2339: Property 'debugLoaders' does not exist on type 'StatsDefaultModuleProps'.

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10330_10330&logFilter=debug&logView=flowAware)
    client/my-sites/stats/features/modules/stats-clicks/stats-clicks.tsx(26,2): error TS2339: Property 'debugLoaders' does not exist on type 'StatsDefaultModuleProps'.

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10331_10331&logFilter=debug&logView=flowAware)
    client/my-sites/stats/features/modules/stats-countries/stats-countries.tsx(26,2): error TS2339: Property 'debugLoaders' does not exist on type 'StatsDefaultModuleProps'.

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10332_10332&logFilter=debug&logView=flowAware)
    client/my-sites/stats/features/modules/stats-referrers/stats-referrers.tsx(26,2): error TS2339: Property 'debugLoaders' does not exist on type 'StatsDefaultModuleProps'.

[02:24:24 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12859528?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=12859528_10333_10333&logFilter=debug&logView=flowAware)
    client/my-sites/stats/features/modules/stats-top-posts/stats-top-posts.tsx(26,2): error TS2339: Property 'debugLoaders' does not exist on type 'StatsDefaultModuleProps'.```

File: client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx [hunk 1]
```diff
@@ -12,45 +13,64 @@ import {
...
-				<StatsCard
+			{ isRequestingData && (
+				<StatsCardSkeleton
+					isLoading={ isRequestingData }
 					className={ className }
+					title={ moduleStrings.title }
+					type={ 2 }
+				/>
+			) }
+			{ ( ( ! isRequestingData && !! data?.length ) || shouldGateStatsModule ) && (
```
Comment (kangzj) [inline client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx:54]: I'm wondering what happens if `data.length` is 0, but `shouldGateStatsModule` is true.

File: client/my-sites/stats/site.jsx [hunk 2]
```diff
@@ -391,6 +391,20 @@ class StatsSite extends Component {
 
 					{ ! isOdysseyStats && <MiniCarousel slug={ slug } isSitePrivate={ isSitePrivate } /> }
 
+					{ isNewStateEnabled && ( // DELETE BEFORE MERGING
```
Comment (kangzj) [inline client/my-sites/stats/site.jsx]: You also mentioned you wanted to remove this before merging 😄

**Comment by grzegorz-cp** [comment]
The type error is due to the manual debug toggle. I'll remove it in the next commit.

File: client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx (hunk 1 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx:54]: If `shouldGateStatsModule` is true it should show the overlay without fetching data.

Comment (kangzj) [reply client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx:54]: `StatsModule` and `StatsCard` would both be rendered -  I didn't really test it but wondering whether that could be a problem 🤔

Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx:54]: Right, that's an issue. I'll check it and update the condition.

**Review by kangzj — APPROVED** [review]
:shipit:

File: client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx (hunk 1 above)
Comment (grzegorz-cp) [reply client/my-sites/stats/features/modules/stats-authors/stats-authors.tsx:54]: It's updated now.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16253596

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @grzegorz-cp for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16253596).

**Pull request merged**